### DISCUSSION TITLE
Welcome to SwissQRBill Discussions!

### DISCUSSION THREAD

**Discussion by manuelbl** [manuelbl/SwissQRBill#57] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below and tell us what you are doing (or trying to do) with QR bills.

**Comment by RogPAS** [comment]
Hallo Manuel, ich wollte wissen, ob Du den SwissQRBill auch in PlentyMarkets implementieren könntest?
Danke und Gruss
Roger

**Comment by manuelbl** [comment]
Hallo Roger

Es gibt keine solchen Pläne. Ich kenn das Produkt gar nicht.

Viele Grüsse
Manuel

**Comment by rnr2lf** [comment]
Hello Manuel,

Is it possible to make the QR-Codes scannable with TWINT ? Or is it planned in a future version?

Thank you and best regards,
Halil

**Reply by manuelbl** [reply]
My limited understanding is for TWINT scannable QR bills is:

1. The billing party needs to contact TWINT and needs to register as a merchant (or something similar). Also see: https://www.twint.ch/en/business-customers/our-solutions/bill-via-twint-qr-code/swiss-qr-code/
2. It is then given QR code string
3. Using the QR code string, an alternative scheme is added to the QR bill.

This library can help with step 3. It will look like this:

```
bill.setAlternativeSchemes(new AlternativeScheme[] {
    new AlternativeScheme("Twint", "twint/light/" + twintQrCodeString)
});
```

Twint should be able to give the exact specification. I haven't test it yet. The description also mentions the integration of the TWINT web interface. I don't really understand that part.

A merchant will likely need to pay a fee for each payment to TWINT.

TWINT also offers a second approach for invoicing. It's also uses a QR code but is not related to the Swiss QR bill standard in anyway. A standard QR code library will likely be sufficient for this case. Also see: https://www.twint.ch/en/business-customers/our-solutions/bill-via-twint-qr-code/

**Reply by rnr2lf** [reply]
Hi Manuel,

Thank you for the fast and precise answer!
What we need is to be able to scan the "normal" Swiss QR-Bill also with TWINT.

I will also ask TWINT how to get this twintQrCodeString.

I will comment the info here in case other people need it.

Thank you again and best regards,
Halil

**Comment by Cousnouf** [comment]
Hi Manuel,
Good job with this library! thanks a lot!

**Comment by uakkaya** [comment]
Hallo Manuel,
vielen Dank für diese Bibliothek. Ich habe evtl. ein Problem festgestellt, vielleicht kannst du dich dazu äußern und mir helfen.
Ich erzeuge erfolgreich den kompletten Zahlschein bisher. Ich möchte jetzt aber einen Zahlschein ohne Referenz erzeugen (bill.setReference(null);), auf der Seite https://www.codecrete.net/qrbill/ ist dieses möglich (so verstehe ich es), lasse ich jedoch in meinem Java Programm das Feld mit null oder einem Leerstring knallt es:
`Caused by: net.codecrete.qrbill.generator.QRBillValidationError: QR bill data is invalid: QR reference is missing; it is mandatory for payments to a QR-IBAN account (qr_ref_missing)
	at net.codecrete.qrbill.generator.QRBill.validateAndGenerate(QRBill.java:219) ~[qrbill-generator-3.0.7.jar:3.0.7]
	at net.codecrete.qrbill.generator.QRBill.generate(QRBill.java:142) ~[qrbill-generator-3.0.7.jar:3.0.7]
	at de.evelopment.p360.vorgVerw.dokumente.model.gen.QRBillGenerator.generate(QRBillGenerator.java:116) ~[classes/:?]`
	
benutze ich den falschen Setter oder muss ich in diesem Fall etwas ganz anderes machen?

Ich danke schon mal im voraus!

Viele Grüße

**Reply by manuelbl** [reply]
Ob die Referenz leer sein kann, hängt von der IBAN ab. IBANs, die an der fünften Stelle eine 3 haben, z.B. "CH44 3199 9123 0008 8901 2", sind so genannte QR-IBANs. Bei diesen muss die Referenz zwingend vorhanden und numerisch sein. Bei anderen IBANs ist die Referenz optional. Ist sie vorhanden, muss sie dem ISO-11649-Format folgen (und also mit "RF" beginnen). Dies ist so vom Standard vorgegeben.

**Reply by uakkaya** [reply]
Oha, das ist mir echt nicht untergekommen in der Doku die ich bekommen habe :(
Ich gehe dann noch mal los und bespreche das. Vielen Dank für die schnelle Antwort.

**Comment by MathiasZaja** [comment]
Hi Manuel,
Are you planning to make your library compatible with PdfBox version 3.0.0 ?
Regards

**Reply by manuelbl** [reply]
Yes, I‘m planning to do so in the near future. Preferably the library should be compatible with both PdfBox 2.x and 3.x.

**Reply by MathiasZaja** [reply]
Thanks for your answer.
Of course for you as a provider of widely used tool you have to offer the compatibilty with both PdfBox.
For myself I adapted your code to PdfBox 3.0.0. There are only some lines to change in PDFCanvas:
a) document = PDDocument.load(Files.newInputStream(path)) -> document = Loader.loadPDF(path.toFile()
b) contentStream.setFont(isBold ? PDType1Font.HELVETICA_BOLD : PDType1Font.HELVETICA,fontSize) ->
  PDFont helveticaBold = new PDType1Font(FontName.HELVETICA_BOLD);
  PDFont helvetica = new PDType1Font(FontName.HELVETICA);
  contentStream.setFont(isBold ? helveticaBold : helvetica, fontSize);
And should be added the 2 librararies: fontbox-3.0.0.jar and pdfbox-io-3.0.0.jar. The io part having been removed from fontbox.
My "migration" is fine with a small exception.
I add "your" bill at the bottom of the last page [PDFCanvas(path2file, PDFCanvas.LAST_PAGE)] of a built and saved pdf which has an image loaded from an URL via PDPageContentStream.drawImage(...)
And the issue is that image is altered in the final document.
Any idea which could help me to find the reason why it happens ?
Regards

**Reply by MathiasZaja** [reply]
PS:
I solved the issue of the altered image when calling canvas.
In the new (? 3.0.7) version of QR Bill Generator canvas.saveAs(path) changed behaviour.
As you mention it: "This method may only be called if the PDF document has been opened or created by this instance. It may not be called if this instance was created by PDFCanvas(PDDocument, int) or PDFCanvas(PDPageContentStream)."
**In practice it means that (unless there is a better way) before calling canvas.saveAs(originalPathToPdfFile) the original PdfFile has to be deleted.**
It took me a while to realise it so it might be helpfull for other folks who will upgrade to this new version of QR Bill Generator.
Regards

**Reply by manuelbl** [reply]
Version 3.0.7 fixed a bug with the PDF graphics context leaking into and out of the QR bill drawing. Possibly, you have been affected by this bug. But it's difficult to tell from your description. However, the behavior of `saveAs()` has not been changed in version 3.0.7.

`saveAs()` is for saving a PDF document that has been created or opened with the Swiss QR bill library. If the document is created or opened with the PDFBox library, it will need to be saved with the PDFBox library as well.

I don't quite understand why you think that `saveAs()` requires to delete the original PDF file. Can you explain this in more detail?

Release 3.2.0 of the Swiss QR Bill library has been released. It is now compatible with both version 2.0 and 3.0 of PDFBox.

**Comment by woddel** [comment]
Hallo Manuel
Wir verwenden das Modul nun seit einiger Zeit.  Vielen Dank dafür!
Eine Frage die nun aufgekommen ist: Wie kann man erweiterte Adressen erfassen?  also mehr als nur VN, NN, Strasse, PLZ, ORT?
Beispiel:
Peter Testname
Alterspflegeheim BeautifulView
Hauptstrasse 111
4000 Irgendwo
Ich hab da keine entsprechenden Felder gefunden in der codecrete.qrbill.generator.Address-Funktion. Ist das überhaupt möglich oder vorgesehen im Schema?
Thanks für Deine Infos dazu! Lukas

**Reply by manuelbl** [reply]
Hallo

Die Library unterstützt keine zusätzlichen Adresszeilen oder Adressfelder, weil es sie im Schweizer Zahlungsverkehr nicht gibt und sie entsprechend nicht im QR-Code codiert werden können. Zudem schreibt der QR-Rechnungsstandard vor, dass auf der QR-Rechnung nur der Text aufgedruckt werden darf, der auch im QR-Code codiert ist.

Das ist natürlich unpraktisch. Es zwingt einen Rechnungssteller faktisch, mit zwei Adressen zu arbeiten: der Postadresse für die Zustellung der Rechnung (oberhalb der eigentlichen QR-Rechnung oder auf einer anderen Seite) und der Adresse für die QR-Rechnung. Aber so ist der Standard.

Ab November 2025 gelten zudem im Schweizer Zahlungsverkehr noch strengere Vorschriften: Es sind nur noch strukturierte Adressen aus Name, Strasse, Hausnummer, PLZ, Ort und Land zulässig. Die unstrukturierten werden nicht mehr akzeptiert.

**Reply by woddel** [reply]
ah supi.  vielen dank für die antwort.  ne, wenn das so sein muss, dann ist das für mich natürlich klar - dann kann ich das auch dem kunden gegenüber argumentieren falls er danach fragt.  beste grüsse, lukas
